test/integration/buffer_accounting_integration_test.cc

```diff
@@ -396,4 +405,202 @@ TEST_P(ProtocolsBufferWatermarksTest, ResettingStreamUnregistersAccount) {
+        ->set_minimum_account_to_track_power_of_two(absl::bit_width(4096u));
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      *bootstrap.mutable_overload_manager() = this->overload_manager_config_;
+    });
+  }
+};
+
+// Run the tests using HTTP2 only since its the only protocol that's fully
+// supported.
+// TODO(kbaichoo): Instantiate with H3 and H1 as well when their buffers are
+// bounded to accounts.
+INSTANTIATE_TEST_SUITE_P(
+    IpVersions, Http2OverloadManagerIntegrationTest,
+    testing::Combine(testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams(
+                         {Http::CodecType::HTTP2}, {FakeHttpConnection::Type::HTTP2})),
+                     testing::Bool()),
+    protocolTestParamsAndBoolToString);
+
+TEST_P(Http2OverloadManagerIntegrationTest,
+       ResetsExpensiveStreamsWhenUpstreamBuffersTakeTooMuchSpaceAndOverloaded) {
+  autonomous_upstream_ = true;
+  autonomous_allow_incomplete_streams_ = true;
+  initializeOverloadManagerInBootstrap(
+      TestUtility::parseYaml<envoy::config::overload::v3::OverloadAction>(R"EOF(
+      name: "envoy.overload_actions.reset_streams"
+      triggers:
+        - name: "envoy.resource_monitors.testonly.fake_resource_monitor"
+          scaled:
+            scaling_threshold: 0.90
+            saturation_threshold: 0.98
+    )EOF"));
+  initialize();
+
+  // Makes us have Envoy's writes to upstream return EAGAIN
+  writev_matcher_->setDestinationPort(fake_upstreams_[0]->localAddress()->ip()->port());
+  writev_matcher_->setWritevReturnsEgain();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto largest_request_response = std::move(sendRequests(1, 4096 * 4, 4096)[0]);
+  auto medium_request_response = std::move(sendRequests(1, 4096 * 2, 4096)[0]);
+  auto smallest_request_response = std::move(sendRequests(1, 4096, 4096)[0]);
+
+  // Wait for requests to come into Envoy.
+  EXPECT_TRUE(buffer_factory_->waitUntilTotalBufferedExceeds(7 * 4096));
+
+  // Set the pressure so the overload action kicks in
+  updateResource(0.95);
+  test_server_->waitForGaugeEq("overload.envoy.overload_actions.reset_streams.scale_percent", 62);
+
+  // Wait for the proxy to notice and take action for the overload by only
+  // resetting the largest stream.
+  if (streamBufferAccounting()) {
+    test_server_->waitForCounterGe("http.config_test.downstream_rq_rx_reset", 1);
+    EXPECT_TRUE(largest_request_response->waitForReset());
+    EXPECT_TRUE(largest_request_response->reset());
+
+    ASSERT_FALSE(medium_request_response->complete());
+  }
+
+  // Increase resource pressure to reset the medium request
+  updateResource(0.96);
+
+  // Wait for the proxy to notice and take action for the overload.
+  if (streamBufferAccounting()) {
+    test_server_->waitForCounterGe("http.config_test.downstream_rq_rx_reset", 2);
+    EXPECT_TRUE(medium_request_response->waitForReset());
```

Comment:
can we check reset reason or response code details here to make sure the accounting works?

Reply:
For reset reason: We get `HTTP::StreamResetReason::RemoteReset` from the endpoint (client). I think that's because it's the catch-all in `ConnectionImpl::onStreamClose`.

https://github.com/envoyproxy/envoy/blob/744b7bfb16273b0496f58382c1e32bf00d29d720/source/common/http/http2/codec_impl.cc#L1082

We could add additional plumbing to better tune this e.g. have the `RST_REASON` sent to downstream be INTERNAL_ERROR (0x2)

as per https://www.rfc-editor.org/rfc/rfc7540.html#section-7

and have the client (e.g. onStreamClose) check for this to guess the reason... though there might be multiple reasons client might send internal error? e.g. Internal Error RST -> Overload might be a mapping, but there might be more?

WDYT?

Reply:
if the client in this test gets remote reset I think that's fine - internal error might be a bit more accurate but I don't mind the catchall. 
The real request here is that we have a way to debug what's going on in production if we see a bunch of streams being reset, and want to verify the cause.  Normally my catchall is "check the stream response code details" but I _think_ the way we reset this through the stream interface we don't have a way to send details, or stick them directly in the stream info.  Alternately there stats which were added somewhere? I'm game for stats or details but I think we should have a way to test Envoy is triggering the path that (reading the test) I believe it's triggering since that lets SRE do the same verification live :-)

Reply:
Added an explicit stat for number of streams the action called resetStream on.